BUG,MAINT: Restructure SingleDeviceMemoryPool and locking

seberg opened the pull request:
**I'll keep this as draft for now, it should work OK but I want to re-think and maybe even roll back a few changes.**

This restructures the single device memory pool, trying to keep it "moderate".  But it does have some significant changes:

1. Arena lifetime is now tied to the chunks which own the arena. This is circular (if the chunk is free and part of the arena) although that is invisible to Python.

   The reason twofold:
   * Prevent cleanup of an arena with active chunks.
   * It is nice to avoid re-entrant/recursive locking. That requires that `free()` can safely put the chunk somewhere that is (effectively) atomic. But I would like that somewhere to live on the `arena` and `arena` discovery _might_ need locking (at least if we may randomly delete the arena).

2. Use atomics for any limit calculations to avoid having to think about which lock to use in what order. (May not matter in the end, but doesn't hurt, I think.)

3. Restructure the internals.  ~I had far more maximal thoughts, but for now keep it slightly more minimal. I did opt for using an `std::map` since it should be good at finding the next "large enough" chunk. I think the removed complexity much outweighs the added need for manual INCREF/DECREF.~
   Changed this to an `std::set` which is also good at finding the next slot and has a `uintptr_t` pointing to the actual chunk as the second part of the key. We can thus immediately find the next free chunk and also find the chunk for merging.
   (To be clear, I think a bucket system can probably be slightly more performant, but unless we do a maximal solution this is just as performance and much simpler.)

This means that the code now definitely doesn't need a re-entrant/recursive mutex anymore.

One other small thing to note maybe: Python containers should be pretty safe (also in free-threading), but iterating them would not be.

seberg committed:
BUG,MAINT: Restructure SingleDeviceMemoryPool and locking

This restructures the single device memory pool, trying to keep it
"moderate".  But it does have some significant changes:

1. Arena lifetime is now tied to the chunks which own the arena.
   This is circular (if the chunk is free and part of the arena)
   although that is invisible to Python.

   The reason twofold:
   * Prevent cleanup of an arena with active chunks.
   * It is nice to avoid re-entrant/recursive locking.
     That requires that `free()` can safely put the chunk somewhere
     that is (effectively) atomic. But I would like that somewhere to
     live on the `arena` and `arena` discovery _might_ need locking
     (at least if we may randomly delete the arena).

2. Use atomics for any limit calculations to avoid having to think
   about which lock to use in what order.
   (May not matter in the end, but doesn't hurt, I think.)

3. Restructure the internals.  I had far more maximal thoughts, but
   for now keep it slightly more minimal.
   I did opt for using an `std::map` since it should be good at
   finding the next "large enough" chunk.
   I think the removed complexity much outweighs the added need for
   manual INCREF/DECREF.

This means that the code now definitely doesn't need a
re-entrant/recursive mutex anymore.

One other small thing to note maybe: Python containers should be
pretty safe (also in free-threading), but iterating them would not be.

seberg committed:
BUG: Fix typo leading to memory leak...

seberg committed:
Fix typo bug and add stress-test (best to just run very long locally maybe)

seberg committed:
Simplify code using std::set (not faster, but a lot simpler)

kumaraditya303 commented on cupy/cuda/memory.pyx:
```diff
@@ -1195,26 +1223,32 @@ cdef class SingleDeviceMemoryPool:
         dict _arenas
+        # NOTE: Never use `.lock()` outside a `with nogil:` statement as it
+        # may deadlock (GIL may be unlocked and another thread also locks).
+        cpp_mutex _arena_mutex
```
if you use cython.pymutex instead of C++ lock then this wouldn't be an issue

seberg committed:
Make `SingleDeviceMemoryPool.free()` internal to remove `in_use` dict

We don't need this dict, calling free directly seems so niche and hidden
to not exist.
(You need to call it via `arr.data.mem.pool().free()`, but then why not
use `arr.data.mem.free()`.  You cannot "disown" a memory object so it
makes no sense to call free without one.

seberg committed:
BUG: Fix ref-counting of weakref fetching and use it once more (maybe meaningless optimization, but had it)

seberg commented on cupy/cuda/memory.pyx:
```diff
@@ -990,176 +1022,177 @@ cdef class PooledMemory(BaseMemory):
+            chunk = self.try_merge_chunk(chunk)
 
+        self.index.insert(index_type(chunk.size, <uintptr_t><void *>chunk))
+        cpython.Py_INCREF(chunk)  # index holds a reference now
```
Yeah, let me change that, it's good habit. I guess I would care more in C and there you should try to use `Py_SETREF()` :).
(In Cython it is hard to get it wrong because what you pass in to `Py_INCREF` *must* be a strong reference!)

seberg committed:
address review comment I forgot about

seberg committed:
Merge branch 'main' into simplify-locking

leofang committed:
Merge branch 'main' into simplify-locking
